### ISSUE TITLE
Kokkos: atomic_add with Kokkos::Serial performance slowdown

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4611]
From https://github.com/trilinos/Trilinos/issues/10015 by @jewatkins :

> ## Bug Report
> 
> @trilinos/kokkos
> ### Description
> 
> I noticed a performance regression when using `Kokkos::atomic_add` on a `Kokkos::Serial` build after the 3.5 promotion (somewhere around 20% slowdown, see [sandialabs/Albany#767](https://github.com/sandialabs/Albany/issues/767)). I'm not sure if this is a bug since this is not its intended use but I thought it's worth reporting just in case others in Trilinos, kokkos-kernels are using atomic_add in this way. In my case, I was able to recover most of the performance by removing its use from the Kokkos::Serial build.
> ### Steps to Reproduce
> 
> I can provide the steps to reproduce if anyone is interested but I would imagine this could be seen with any atomic_add usage in a Kokkos::Serial build before/after the kokkos promotion. I saw it in a larger case w/ intel and a smaller case w/ clang. I can also provide data if needed.
> 
> I can also escalate this to a kokkos issue if this is indeed a bug. Otherwise, I can close this.

**Comment by ndellingwood** [comment]
I'm curious if the updates to introduce desul atomics maybe disabled or negated changes related to #1426 ?

**Comment by crtrott** [comment]
yeah. it probably did. We can fix that ...

**Comment by crtrott** [comment]
We have this now in develop @jewatkins can you check this?

**Comment by jewatkins** [comment]
Yeah I'll give it a try

**Comment by jewatkins** [comment]
I get the following error when building Albany with Trilinos/develop + kokkos/develop:
```
CMake Error at /ascldap/users/jwatkin/LandIce/Trilinos.fork/build-blake-intel-serial-kokkos-snapshot/install/lib/cmake/Kokkos/KokkosConfig.cmake:172 (IF):
  if given arguments:

    "Kokkos_ENABLE_THREADS" "AND" "NOT" "DEPRECATED_CODE_3" "IN_LIST" "Kokkos_OPTIONS"

  Unknown arguments specified
```
Any ideas?

**Comment by dalg24** [comment]
What version of CMake are you building Albany with when you get this error?

**Comment by jewatkins** [comment]
cmake/3.19.3

**Comment by dalg24** [comment]
Would you share the relevant lines in your `/ascldap/users/jwatkin/LandIce/Trilinos.fork/build-blake-intel-serial-kokkos-snapshot/install/lib/cmake/Kokkos/KokkosConfig.cmake`

**Comment by jewatkins** [comment]
```
IF(Kokkos_ENABLE_THREADS AND NOT DEPRECATED_CODE_3 IN_LIST Kokkos_OPTIONS)  # for backward compatibility
  SET(Kokkos_ENABLE_PTHREAD ON)
  LIST(APPEND Kokkos_DEVICES PTHREAD)
ENDIF()
```
It seems like the only thing that wouldn't be defined is `Kokkos_ENABLE_THREADS` but not sure. I tried including it during Trilinos configure but I received the same error.

**Comment by dalg24** [comment]
```
SET(Kokkos_DEVICES <????>) # <-- What is enabled?
SET(Kokkos_OPTIONS <????>) # <-- here too
<snip>
FOREACH(DEV ${Kokkos_DEVICES}) # <-- Do you have that loop
  SET(Kokkos_ENABLE_${DEV} ON)
ENDFOREACH()

**Comment by jewatkins** [comment]
```
SET(Kokkos_DEVICES SERIAL)
SET(Kokkos_OPTIONS DEPRECATED_CODE_3;DEPRECATION_WARNINGS;LAUNCH_COMPILER;IMPL_DESUL_ATOMICS)
```
Yes the loop is there.

**Comment by dalg24** [comment]
OK I was able to reproduce with
```CMake
cmake_minimum_required(VERSION 2.8.12)
project(MyProj)
#cmake_policy(SET CMP0057 NEW)
set(MYLIST FOO BAR BAZ)
if(NOTDEFINED AND NOT FOO IN_LIST MYLIST)
  message(STATUS "I <3 Kokkos")
endif()
```
Either set the CMP0057 policy to NEW or require a CMake version >=3.3

**Comment by jewatkins** [comment]
I should have scrolled up a bit more to see the warning...

**Comment by jewatkins** [comment]
Looks good to me:
```
trilinos/kokkos w/o atomic:
1: |   |   |   |   |   |   |   Phalanx: Evaluator 70: [<Jacobian>] Scatter Velocity Residual<Jacobian>: 0.285558 - 25.0763% [11] {min=0.269294, max=0.302525, std dev=0.00957151}
trilinos/kokkos w/ atomic:
1: |   |   |   |   |   |   |   Phalanx: Evaluator 70: [<Jacobian>] Scatter Velocity Residual<Jacobian>: 0.446014 - 33.9484% [11] {min=0.420083, max=0.498198, std dev=0.0144717}
kokkos/develop w/ atomic:
1: |   |   |   |   |   |   |   Phalanx: Evaluator 70: [<Jacobian>] Scatter Velocity Residual<Jacobian>: 0.292559 - 25.6001% [11] {min=0.278357, max=0.312617, std dev=0.0093325}
```

**Comment by PhilMiller** [comment]
Problem addressed by the time 3.6 was released
